src/host/linalg.jl

```diff
@@ -197,33 +197,39 @@ LinearAlgebra.lmul!(a::Number, B::AbstractGPUArray) = generic_lmul!(a, B)
 LinearAlgebra.permutedims!(dest::AbstractGPUArray, src::AbstractGPUArray, perm) =
     permutedims!(dest, src, Tuple(perm))
 
+@inline @generated function permute_linearindex(size::NTuple{N,T}, l::T, strides::NTuple{N,T}) where {N,T}
+    quote
+        l -= one(T)
+        res = one(T)
+        Base.Cartesian.@nexprs $(N-1) i->begin
+            LLVM.Interop.assume(size[i] > 0)
+            @inbounds l, s = divrem(l, size[i])
+            @inbounds res += s * strides[i]
+        end
+        return @inbounds res + strides[N] * l
+    end
+end
 function LinearAlgebra.permutedims!(dest::AbstractGPUArray, src::AbstractGPUArray,
                                     perm::NTuple{N}) where N
+    @assert length(src) <= typemax(UInt32)
     Base.checkdims_perm(dest, src, perm)
-
-    # get the new strides of destination tensor
     dest_strides = ntuple(k->k==1 ? 1 : prod(i->size(dest, i), 1:k-1), N)
-    dest_strides_perm = ntuple(i->dest_strides[findfirst(==(i), perm)], N)
-
-    function permutedims_kernel(ctx, dest, src, dest_strides_perm)
-        # find the cartesian index in source tensor
-        LI = @linearidx src
-        I = @inbounds CartesianIndices(src)[LI]
-
-        # the corresponding linear index in the destination tensor
-        dest_index = map_index(I.I, dest_strides_perm)
+    dest_strides_perm = ntuple(i->UInt32(dest_strides[findfirst(==(i), perm)]), N)
```

Comment:
What with arrays whose indices don't fit in UInt32? Wouldn't it be better to have two kernels (or a generically typed one), figure out on the hose which one we can use and dispatch/specialize appropriately?